ghcide/src/Development/IDE/Plugin/ChangeTypeAction.hs

```diff
@@ -0,0 +1,130 @@
+
+-- | Create a CodeAction from a Diagnostic
+generateAction :: SigName p => Uri -> [LHsDecl (GhcPass p)] -> Diagnostic -> Maybe (Command |? CodeAction)
+generateAction uri decls diag = diagnosticToChangeSig uri decls diag >>= changeSigToCodeAction
+
+-- | Convert a diagnostic into a ChangeSignature and add the proper SrcSpan
+diagnosticToChangeSig :: SigName p => Uri -> [LHsDecl (GhcPass p)] -> Diagnostic -> Maybe ChangeSignature
+diagnosticToChangeSig uri decls diag = addSrcSpan decls <$> matchingDiagnostic uri diag
+
+-- | If a diagnostic has the proper message create a ChangeSignature from it
+matchingDiagnostic :: Uri -> Diagnostic -> Maybe ChangeSignature
+matchingDiagnostic uri diag@Diagnostic{_message} = unwrapMatch $ _message =~ expectedMessage
+    where
+        expectedMessage = "Expected type: (.+)\n +Actual type: (.+)\n.*\n +In an equation for ‘(.+)’" :: Text
+        unwrapMatch :: (Text, Text, Text, [Text]) -> Maybe ChangeSignature
+        unwrapMatch (_, _, _, [exp, act, name]) = Just $ ChangeSignature exp act name Nothing diag uri
+        unwrapMatch _                           = Nothing
+
+-- | Given a String with the name of a declaration, find that declarations type signature location
+-- This is a modified version of functions found in Development.IDE.Plugin.CodeAction
+findSigLocOfStringDecl :: SigName p => [LHsDecl (GhcPass p)] -> String -> Maybe SrcSpan
+findSigLocOfStringDecl decls declName =
+  listToMaybe
+    [ locA srcSpan
+      | L srcSpan (SigD _ (TypeSig _ idsSig _)) <- decls,
+        any ((==) declName . occNameString . occName . unLoc) idsSig
+    ]
+
+-- | Update a ChangeSignature to potentially populate `declSrcSpan`
+addSrcSpan :: SigName p => [LHsDecl (GhcPass p)] -> ChangeSignature -> ChangeSignature
+addSrcSpan _ self@(ChangeSignature _ _ _ (Just _) _ _) = self
+addSrcSpan decls chgSig@ChangeSignature{..} = chgSig { declSrcSpan = findSigLocOfStringDecl decls (T.unpack declName) }
+
+changeSigToCodeAction :: ChangeSignature -> Maybe (Command |? CodeAction)
+-- Does not generate a Code action if declSrcSpan is Nothing
+changeSigToCodeAction ChangeSignature{..} = declSrcSpan *> Just (InR CodeAction { _title       = mkChangeSigTitle declName actualType
+                                                                                , _kind        = Just CodeActionQuickFix
+                                                                                , _diagnostics = Just $ List [diagnostic]
+                                                                                , _isPreferred = Nothing
+                                                                                , _disabled    = Nothing
+                                                                                -- This CAN but probably never will be Nothing
+                                                                                , _edit        = mkChangeSigEdit uri declSrcSpan (mkNewSignature declName actualType)
```

Comment:
It seems a little awkward to pass in a `Maybe SrcSpan` here when you should know it's there by now. Why not write this with a do block?
```
do 
  srcSpan <- declSrcSpan
  pure $ InR CodeAction { ... , mkChangeSigEdit uri srcSpan ... }
```
and then `mkChangeEdit` can take a non-`Maybe` `SrcSpan`.